lib/muser_ctx.c

```diff
@@ -499,70 +499,78 @@ handle_device_get_info(lm_ctx_t *lm_ctx, uint32_t size,
+ *          the number of DMA regions in @dma_regions.
+ * @dma_regions: memory that contains the DMA regions to be mapped/unmapped
+ *
+ * @returns 0 on success, -errno on failure.
+ */
 int
 handle_dma_map_or_unmap(lm_ctx_t *lm_ctx, uint32_t size, bool map,
                         int *fds, int nr_fds,
                         struct vfio_user_dma_region *dma_regions)
 {
     int nr_dma_regions;
-    int fdi = 0;
     int ret, i;
 
     assert(lm_ctx != NULL);
+    assert(fds != NULL);
+
+    if (lm_ctx->dma == NULL) {
+        return 0;
+    }
 
     if (size % sizeof(struct vfio_user_dma_region) != 0) {
         lm_log(lm_ctx, LM_ERR, "bad size of DMA regions %d", size);
         return -EINVAL;
     }
 
-    if (lm_ctx->dma == NULL) {
-        return 0;
-    }
-
     nr_dma_regions = (int)(size / sizeof(struct vfio_user_dma_region));
+    if (map && nr_fds > 0 && nr_dma_regions != nr_fds) {
+        lm_log(lm_ctx, LM_ERR, "expected %d fds but got %d instead",
+               nr_dma_regions, nr_fds);
+        return -EINVAL;
+    }
 
     for (i = 0; i < nr_dma_regions; i++) {
         if (map) {
-            int fd;
-
-            /*
-             * FIXME: need a dma controller that allows non-fd region.
-             */
-            if (dma_regions[i].flags != VFIO_USER_F_DMA_REGION_MAPPABLE) {
-                lm_log(lm_ctx, LM_INF,
-                       "FIXME: ignored non-mappable DMA region "
-                       "%#lx-%#lx offset=%#lx",
-                       dma_regions[i].addr,
-                       dma_regions[i].addr + dma_regions[i].size - 1,
-                       dma_regions[i].offset);
-                continue;
-            }
-
-            if (fdi >= nr_fds) {
-                lm_log(lm_ctx, LM_ERR, "missing fd for mappable region %d", i);
-                return -EINVAL;
+            if (dma_regions[i].flags == VFIO_USER_F_DMA_REGION_MAPPABLE) {
+                if (nr_fds == 0) {
+                    return -EINVAL;
+                }
+            } else {
+                if (nr_fds != 0) {
+                    return -EINVAL;
+                }
+                fds[i] = -1;
             }
 
-            fd = fds[fdi++];
-
             ret = dma_controller_add_region(lm_ctx->dma,
                                             dma_regions[i].addr,
                                             dma_regions[i].size,
-                                            fd,
+                                            fds[i],
                                             dma_regions[i].offset);
             if (ret < 0) {
                 lm_log(lm_ctx, LM_INF,
                        "failed to add DMA region %#lx-%#lx offset=%#lx fd=%d: %s",
                        dma_regions[i].addr,
                        dma_regions[i].addr + dma_regions[i].size - 1,
-                       dma_regions[i].offset, fd,
+                       dma_regions[i].offset, fds[i],
```

Comment:
This looks like it broke my fix. Consider two regions being passed in:
region[0] !mappable
region[1] mappable

When we get to region[1], i == 1, but the fd for it is in fd[0] (which you over-wrote with -1)

Reply:
Ah yes, though we don't allow it because of the check we do earlier:
```
if (map && nr_fds > 0 && nr_dma_regions != nr_fds) {
        lm_log(lm_ctx, LM_ERR, "expected %d fds but got %d instead",
               nr_dma_regions, nr_fds);
        return -EINVAL;
    }
```
However this check is wrong since we should allow mappable and non-mappable regions in the same message. Let me fix this.

Reply:
@jlevon fixed, please take another look.